Wreturn-type-c-linkage warning from Clang

The demo works but emits a warning.

```console
% cargo run --manifest-path demo-rs/Cargo.toml
   Compiling proc-macro2 v1.0.8
   Compiling unicode-xid v0.2.0
   Compiling cc v1.0.50
   Compiling syn v1.0.14
   Compiling unicode-segmentation v1.6.0
   Compiling anyhow v1.0.26
   Compiling unicode-width v0.1.7
   Compiling termcolor v1.1.0
   Compiling codespan v0.7.0
   Compiling codespan-reporting v0.7.0
   Compiling link-cplusplus v1.0.1
   Compiling cxx v0.1.2
   Compiling quote v1.0.2
   Compiling thiserror-impl v1.0.9
   Compiling cxxbridge-macro v0.1.2
   Compiling thiserror v1.0.9
   Compiling cxxbridge-demo v0.0.0
warning: target/debug/build/cxxbridge-demo-5144ad694551d3cf/out/demo-rs/src/main.rs.cc:94:20: warning: 'org$rust$cxxbridge01$get_name' has C-linkage specified, but returns user-defined type 'const std::string &' (aka 'const basic_string<char, char_traits<char>, allocator<char> > &') which is incompatible with C [-Wreturn-type-c-linkage]
warning: const std::string &org$rust$cxxbridge01$get_name(const ThingC &thing) noexcept {
warning:                    ^
warning: 1 warning generated.
    Finished dev [unoptimized + debuginfo] target(s) in 16.65s
     Running `target/debug/cxxbridge-demo`
this is a demo of cxx::bridge
called back with r=333
done with ThingC
```

--- a/gen/write.rs
+++ b/gen/write.rs
@@ -211,8 +211,11 @@ fn write_cxx_function_shim(out: &mut OutFile, efn: &ExternFn, types: &Types) {
         write!(out, "new (return$) ");
         write_type(out, efn.ret.as_ref().unwrap());
         write!(out, "(");
-    } else if efn.ret.is_some() {
+    } else if let Some(ret) = &efn.ret {
         write!(out, "return ");
+        if let Type::Ref(_) = ret {
+            write!(out, "&");
+        }
     }
     write!(out, "{}$(", efn.ident);
     for (i, arg) in efn.args.iter().enumerate() {
@@ -298,8 +301,11 @@ fn write_rust_function_shim(out: &mut OutFile, efn: &ExternFn, types: &Types) {
             write_type(out, efn.ret.as_ref().unwrap());
             writeln!(out, ")];");
             write!(out, "  ");
-        } else if efn.ret.is_some() {
+        } else if let Some(ret) = &efn.ret {
             write!(out, "return ");
+            if let Type::Ref(_) = ret {
+                write!(out, "*");
+            }
         }
         for name in out.namespace.clone() {
             write!(out, "{}$", name);
@@ -351,6 +357,13 @@ fn write_extern_return_type(out: &mut OutFile, ty: &Option<Type>, types: &Types)
         Some(Type::RustBox(ty)) | Some(Type::UniquePtr(ty)) => {
             write_type_space(out, &ty.inner);
             write!(out, "*");
+        }
+        Some(Type::Ref(ty)) => {
+            if ty.mutability.is_none() {
+                write!(out, "const ");
+            }
+            write_type(out, &ty.inner);
+            write!(out, " *");
         }
         Some(Type::Str(_)) => write!(out, "cxxbridge::RustStr::Repr "),
         Some(ty) if types.needs_indirect_abi(ty) => write!(out, "void "),

